Subject: [GTI pull] hrtimer fixes

From: Thomas Gleixner <tglx@linutronix.de>

Linus,


please pull hrtimer fixes from:

  ssh://master.kernel.org/pub/scm/linux/kernel/git/tglx/linux-2.6-hrt.git master

This fixes a long standing hrtimer reprogramming bug.

Thanks,
	tglx
---

 kernel/hrtimer.c |   15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

Index: linux-2.6.24.7-rt27/kernel/hrtimer.c
===================================================================
--- linux-2.6.24.7-rt27.orig/kernel/hrtimer.c	2009-02-08 00:00:39.000000000 -0500
+++ linux-2.6.24.7-rt27/kernel/hrtimer.c	2009-02-08 00:00:41.000000000 -0500
@@ -1172,8 +1172,19 @@ static void run_hrtimer_softirq(struct s
 			 * If the timer was rearmed on another CPU, reprogram
 			 * the event device.
 			 */
-			if (timer->base->first == &timer->node)
-				hrtimer_reprogram(timer, timer->base);
+			struct hrtimer_clock_base *base = timer->base;
+
+			if (base->first == &timer->node &&
+			    hrtimer_reprogram(timer, base)) {
+				/*
+				 * Timer is expired. Thus move it from tree to
+				 * pending list again.
+				 */
+				__remove_hrtimer(timer, base,
+						 HRTIMER_STATE_PENDING, 0);
+				list_add_tail(&timer->cb_entry,
+					      &base->cpu_base->cb_pending);
+			}
 		}
 	}
 	spin_unlock_irq(&cpu_base->lock);
